[WebGPU] Infinite loop prevention does not need to be applied to concrete initialization

mwyrzykowski opened the pull request:
#### 4d4b88e9f8093076b4ae3960fcf7c17d70968e22
<pre>
[WebGPU] Infinite loop prevention does not need to be applied to concrete initialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=292959">https://bugs.webkit.org/show_bug.cgi?id=292959</a>
<a href="https://rdar.apple.com/151196355">rdar://151196355</a>

Reviewed by Tadeu Zagallo.

Preventing against UB due to infinite loops is not needed for for loops
we emit ourselves to initialize data members.

* LayoutTests/fast/webgpu/nocrash/fuzz-292959-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-292959.html: Added.
Add regression test.

* Source/WebGPU/WGSL/AST/ASTForStatement.h:
No need to apply UB prevention to our own for loops.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
Disable optimization on __workgroup_uniform_load so the test passes on Apple9 devices.

(WGSL::Metal::FunctionDefinitionWriter::visit):

Canonical link: <a href="https://commits.webkit.org/295250@main">https://commits.webkit.org/295250@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e32b1b0d9e3100319cdd67b3124d3550b33a06b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104477 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24187 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109684 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55148 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24568 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32731 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79328 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19111 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94301 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59655 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18899 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12373 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54517 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88599 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12431 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112069 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31638 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23302 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88370 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32002 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90534 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88036 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22432 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32939 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10713 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/26115 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31565 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36905 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31357 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34696 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32917 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -423,7 +423,7 @@ void FunctionDefinitionWriter::emitNecessaryHelpers()
     if (m_shaderModule.usesWorkgroupUniformLoad()) {
         m_body.append(m_indent, "template<typename T>\n"_s,
-            m_indent, "static T __workgroup_uniform_load(threadgroup T* const ptr)\n"_s,
+            m_indent, (metalAppleGPUFamily() >= 9 ? "[[clang::optnone]] "_s : ""_s), "static T __workgroup_uniform_load(threadgroup T* const ptr)\n"_s,
```
This depends on `https://github.com/WebKit/WebKit/pull/45322` for `metalAppleGPUFamily` definition

mwyrzykowski replied:
When I look at the generated code, it appears this is a shader validation bug. So once https://github.com/WebKit/WebKit/pull/45322 lands we should conditionalize this for SV-only just so we don't see these false positives.

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTForStatement.h:
```diff
@@ -46,12 +47,14 @@ class ForStatement final : public Statement {
         , m_test(test)
         , m_update(update)
         , m_body(WTFMove(body))
+        , m_isInternallyGenerated(true)
```
looks like this will always be true since I believe the parser uses this same constructor, no?

mwyrzykowski replied:
oh yes the other loop in the test is a `while` which is why it was still getting the UB prevention, I will fix that, thank you for noticing :)

mwyrzykowski committed:
[WebGPU] Infinite loop prevention does not need to be applied to concrete initialization
https://bugs.webkit.org/show_bug.cgi?id=292959
rdar://151196355

Reviewed by Tadeu Zagallo.

Preventing against UB due to infinite loops is not needed for for loops
we emit ourselves to initialize data members.

* LayoutTests/fast/webgpu/nocrash/fuzz-292959-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-292959.html: Added.
Add regression test.

* Source/WebGPU/WGSL/AST/ASTForStatement.h:
No need to apply UB prevention to our own for loops.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
Disable optimization on __workgroup_uniform_load so the test passes on Apple9 devices.

(WGSL::Metal::FunctionDefinitionWriter::visit):

Canonical link: https://commits.webkit.org/295250@main
